### DISCUSSION TITLE
ValueError: 'c' argument has 26 elements, which is inconsistent with 'x' and 'y' with size 29

### DISCUSSION THREAD

**Discussion by goutham-kumar** [networkx/networkx#5100] [Q&A] [answered]
I see the below error while creating a drawing.
`ValueError: 'c' argument has 26 elements, which is inconsistent with 'x' and 'y' with size 29`

I have been trying to find the reason behind the exception but couldnt point, could someone please help me out. Have been working on this since a week.

Few more info,
we are migrating to py3 recently, I see no issues causing while we were using py2. Dont understand what made this error popup. Code is intact.

**Comment by rossbar** [comment]
When reporting issues, it's generally helpful to include a *minimal reproducing example* that results in the traceback. Fortunately, in this case the exception is quite clear and gives us a good sense of what the likely problem is: I would guess that you are drawing nodes and specifying an explicit color for each node, but there is a mismatch between the size of "node_colors" and the number of nodes. For example, I can reproduce the error you're seeing as follows:

```python
>>> import networkx as nx
>>> G = nx.path_graph(29)
>>> nx.draw_networkx_nodes(G, pos=nx.spring_layout(G), node_color=["k"]*26)
Traceback (most recent call last)
   ...
ValueError: 'c' argument has 26 elements, which is inconsistent with 'x' and 'y' with size 29.
```

So - I would guess that the problem is that there is a mismatch between the number of nodes and the number of node colors you are supplying when drawing.

**Comment by goutham-kumar** [comment]
Thankyou @rossbar for answering my question quickly.

Here is what I use
```
g.add_node(deviceID, label=nlabel, color=nodecolor)
nlabels = networkx.get_node_attributes(g,"label")
ncolors = networkx.get_node_attributes(g,"color").values()
networkx.draw(g, with_labels=True,labels=nlabels,node_color=ncolors)
```
here label is the name of device and color is the intended color.
Not sure what is the problem here, could you please guide.

Thanks in advance

**Comment by dschult** [comment, marked as answer]
Can you verify that the length of `ncolors` is 26 while `len(g)` is 29?
The key to debugging this is to not trust your assumptions. Make a short bit of code (like what you've provided but that will actually work -- you don't actually add any nodes above) that creates the error. Then check `len(g)` and `ncolors` and print the dictionary whose values create `ncolors` and make sure the nodes are all there, etc.
Something is not what you think it is....  That is what is causing the error.

**Reply by goutham-kumar** [reply]
Yes, `ncolors ` is 26 & `len(g)` is 29, but I dont understand why this difference.

Though I am new to networkx, with the code running I understand add_node runs for 26 deviceIDs and I am assigning nodecolor for same devices. Now the question is why that I am getting `ncolors` as 26 but `len(g)` as 29. 

```
for dev in devices:
    if dev.deviceType == "router (IOS)" or dev.deviceType == "router (Other)":
        nodecolor = "#ff8000"
    elif "switch" in dev.deviceType:
        nodecolor = "#0040ff"
    elif dev.deviceType == "accesspoint":
        nodecolor = "#00ff40"
    elif dev.deviceType == "wlc":
        nodecolor = "#00ff80"
    else:
        nodecolor = "#ff0000"
    if dev.deviceName != "":
        nlabel= dev.deviceName + "\n" + dev.deviceModel
    else:
        nlabel= dev.deviceIP + "\n" + dev.deviceModel
    
    g.add_node(dev.deviceID, label=nlabel, color=nodecolor)
    nlabels = networkx.get_node_attributes(g,"label")
    ncolors = networkx.get_node_attributes(g,"color").values()
    networkx.draw(g, with_labels=True,labels=nlabels,node_color=ncolors)
```
Technically speaking both `ncolors` & `len(g)` should be 26 isnt it.

**Reply by MridulS** [reply]
> with the code running I understand add_node runs for 26 deviceIDs and I am assigning nodecolor for same devices. Now the question is why that I am getting ncolors as 26 but len(g) as 29.

Could you try creating an empty graph just before this loop is run? something like
```python
g = networkx.Graph()
for dev in devices:
...
...
```
If the length of the container `devices` is 26, the graph `g` shouldn't contain more than 26 nodes according to this code snippet (assuming all `deviceID` are unique).

**Reply by goutham-kumar** [reply]
Hi @MridulS,
I found out the cuilprit
`g.add_edge(dev.deviceID, neighbor, label=label)`

`neighbor` is the neighboring deviceID.

Tried to print list(g) post the above line, which gave me the node count as 29.


Is it mandatory to to use `g.add_edge` post `g.add_node`.

**Reply by goutham-kumar** [reply]
missed to mention,
Same set of code was working fine with py2, now that we are migrating to py3.

**Reply by MridulS** [reply]
> Is it mandatory to to use g.add_edge post g.add_node

No, it is only required if you want to add an edge.

But `g.add_edge` will add a node to the graph if it doesn't exist. So it looks like there is `neighbor` deviceID here which isn't present in `devices`. I would suggest to check the output of
```python
for dev in devices:
    print(dev.deviceID)
```
and 
```
print(g.nodes())
```
manually. There is a node somewhere which is not in the list (or iterator) `devices` and shows up in `neighbor`.

> Same set of code was working fine with py2, now that we are migrating to py3.

I remember seeing this error a long time back, this behaviour was probably a bug that was fixed.

**Reply by goutham-kumar** [reply]
I did checked the nodes, 3 devicesIDs were not part of the `list(g)` whereas those got added while creating an edge.

Now that I am checking `if list(g) < len(ncolors )`, if true I am adding them to `ncolors` dict with some color. This is how I resolved this issue.

Thanks for your help, I was trying all the way to understand the root cause of the issue, you people helped to find it.
